### PR TITLE
Inlining sync.cpp with latest upstream, first step.

### PR DESCRIPTION
**Pull request by furszy**
Part of a group of updates that will be doing from upstream to get up-to-date with latest sync.h/cpp sources.

This first PR involves:

* sync.h/cpp code refactor + reorg.
* threadnames.h/cpp file creation + connection (where "connection" means that all of the previous `RenameThread` method calls where modified to use the new `util::ThreadRename` method).

### CODE REVIEW

File: src/util/threadnames.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+//! internal name.
+static void SetThreadName(const char* name)
+{
+#if defined(PR_SET_NAME)
+    // Only the first 15 characters are used (16 - NUL terminator)
+    ::prctl(PR_SET_NAME, name, 0, 0, 0);
+#elif (defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__DragonFly__))
+    pthread_set_name_np(pthread_self(), name);
+#elif defined(MAC_OSX)
+    pthread_setname_np(name);
```
Comment (random-zebra) [inline src/util/threadnames.cpp]: is the check for XCode 10.6-and-later no longer needed here?

Comment (furszy) [reply src/util/threadnames.cpp]: yep,  [#5366](https://github.com/bitcoin/bitcoin/pull/5366)

**Comment by Fuzzbawls** [comment]
needs rebase

**Comment by furszy** [comment]
rebased.

**Pull request merged**